File: tests/fixtures/check_then_act/py/own_method_recheck_fixture.py
# A recheck through the object's own method cannot be told from one that reads an unrelated flag, so
# the window stays listed for the grader to clear.
class Service:
    def __init__(self, repo, graph):
        self.repo, self.graph = repo, graph

    async def apply(self, item_id):
        item = self.repo.get(item_id)
        if item.status != 'pending':  # expect: check
            return item
        await self.graph.project(item_id)  # expect: gap await
        after = self.stored(item_id)
        if after.status != 'pending':
            return after
        self.repo.mark_applied(item_id)  # expect: write

    def stored(self, item_id):
        return self.repo.get(item_id)

File: tests/fixtures/check_then_act/ts/own_method_recheck_fixture.ts
// The state is read again after the await through the object's own helper, which may read any store.
export class Service {
  constructor(private readonly repo: any, private readonly graph: any) {}

  async apply(id: string) {
    const item = this.repo.get(id);
    if (item.status !== 'pending') return item;
    await this.graph.project(id);
    const after = this.stored(id);
    if (after.status !== 'pending') return after;
    this.repo.markApplied(id);
  }

  private stored(id: string) {
    return this.repo.get(id);
  }
}
